pysmt/typing.py

```diff
@@ -219,11 +220,21 @@ class _FunctionType(PySMTType):
     This class should not be instantiated directly, but the factory
     method FunctionType should be used instead.
     """
+
+    _instances = {}
+
     def __init__(self, return_type, param_types):
-        PySMTType.__init__(self, type_id = 4)
+        PySMTType.__init__(self)
         self._return_type = return_type
-        self._param_types = param_types
+        self._param_types = tuple(param_types)
         self._hash = hash(return_type) + sum(hash(p) for p in param_types)
+        # Note:
+        # To simplify handling of types, we need to treat FunctionType
+        # in a special way. We do so by setting its arity to a
+        # negative value and creating an args that is the concatenation of
+        # return_type and param_types. In this way, we can iterate on args.
+        self.arity = -1
+        self.args = (self._return_type,) + self.param_types
```

Comment:
This is quite ugly, why can't we set the arity to len(self.param_types) and the return type as either an additional field or an extra cell of args?